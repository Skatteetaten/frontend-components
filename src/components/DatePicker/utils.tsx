export const DEFAULTFORMATDATE = (date: Date | null | undefined): string => {
  return date
    ? Intl.DateTimeFormat('no', {
        year: 'numeric',
        month: '2-digit',
        day: 'numeric',
      }).format(date)
    : '';
};

export const DEFAULTPARSEDATEFROMSTRING = (date: string): Date | null => {
  try {
    if (date && date.length === 10) {
      const day: number = parseInt(date.substring(0, 2));
      const month: number = parseInt(date.substring(3, 5));
      const year: number = parseInt(date.substring(6, 10));
      const formattedDate: Date = new Date(Date.UTC(year, month - 1, day));
      return formattedDate;
    }
  } catch (error) {
    console.log(error);
  }
  return null;
};

export const monthsForLocale = (
  localeName = 'no',
  monthFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { month: monthFormat })
    .format;
  const months: string[] = [];
  for (let month = 0; month < 12; month++) {
    months.push(format(new Date(Date.UTC(2020, month))));
  }
  return months;
};

export const weekdaysForLocale = (
  localeName = 'no',
  dayFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { weekday: dayFormat })
    .format;
  const weekdays: string[] = [];
  for (let day = 0; day < 7; day++) {
    weekdays.push(format(new Date(Date.UTC(2020, 7, day + 2)))); // add 2 days to let the week start on Sunday
  }
  return weekdays;
};
