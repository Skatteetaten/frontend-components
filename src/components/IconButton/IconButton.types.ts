import { ButtonProps } from '../Button';

export interface IconButtonProps extends ButtonProps {
  /** Om sirkel skal vises eller ikke. Sirkel vil typisk benyttes når ikon fra material-design ikke har egen sirkel  */
  circle?: boolean;
  /** Ikon som skal vises foran teksten på knappen */
  icon?: string;
  /** Beskrivelse av hva knappen gjør (f.eks. til skjermlesere) */
  title?: string;
  /** Gjør knappen til hovedknapp med fylt farge */
  buttonSize?: 'default' | 'xSmall' | 'small' | 'medium' | 'large' | 'xLarge';
  /**
   * Fjerne Prop
   * @ignore
   */
  buttonStyle?: ButtonProps['buttonStyle'];
}
