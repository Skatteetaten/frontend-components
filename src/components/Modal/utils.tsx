import { useEffect } from 'react';

export const createModalDomPlacement = (
  setFlag: (a: boolean) => void,
  shadowRootNode: Document | ShadowRoot | undefined
) => {
  if (!shadowRootNode) {
    setFlag(true);
    return;
  }

  const styleInjectorWrapperId = 'style-injector';
  const modalWrapperId = 'modal-wrapper';

  if (shadowRootNode.getElementById(modalWrapperId)) {
    setFlag(true);
  } else {
    const modalWrapper = document.createElement('div');
    modalWrapper.id = modalWrapperId;
    const styleInjectorWrapper = shadowRootNode.getElementById(
      styleInjectorWrapperId
    );
    const parent = styleInjectorWrapper ? styleInjectorWrapper : shadowRootNode;
    parent.appendChild(modalWrapper);
    setFlag(true);
  }
};

export const useEscOnPress = (
  ref: HTMLDivElement | null,
  onClose?: (ref: HTMLDivElement) => void
): void => {
  useEffect(() => {
    const onKeyDown = (event: KeyboardEvent): void => {
      if (ref == null) {
        return;
      }
      if (event.code === 'Escape' && onClose) {
        onClose(ref);
      }
    };

    if (ref) {
      window.addEventListener('keydown', onKeyDown as EventListener, false);
    }
    return function cleanup(): void {
      if (window) {
        window.removeEventListener('keydown', onKeyDown as EventListener);
      }
    };
  }, [ref, onClose]);
};
