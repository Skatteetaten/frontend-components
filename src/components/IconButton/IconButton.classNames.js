import { mergeStyles } from '@uifabric/merge-styles';
import { getTheme } from '@uifabric/styling';
import { FontSizes } from '../utils/fonts';
import { getFocusStyle } from '../utils/getFocusStyle';

function getTypeColor(props) {
  switch (props.buttonType) {
    case 'small':
      return {
        width: '35px',
        height: '35px'
      };
    case 'medium':
      return {
        width: '40px',
        height: '40px'
      };
    case 'large':
      return {
        width: '50px',
        height: '50px'
      };
    case 'xLarge':
      return {
        width: '60px',
        height: '60px'
      };
    default:
      return {};
  }
}

function getCircleStyles(props) {
  const { palette } = getTheme();

  if (props.circle === true) {
    return {
      border: 'solid',
      borderWidth: '3px',
      borderColor: palette.skeColor.blue
    };
  } else {
    return {
      borderWidth: 0
    };
  }
}

function getCircleHoverStyles(props) {
  const { palette } = getTheme();

  if (props.circle === true) {
    return {
      borderWidth: '3px',
      background: palette.skeColor.lightBlue,
      color: palette.skeColor.blue
    };
  } else {
    return {
      color: palette.skeColor.blue,
      borderWidth: '3px',
      borderColor: palette.skeColor.blue,
      background: palette.skeColor.lightBlue
    };
  }
}

function getCircleFocusrStyles(props) {
  const { palette } = getTheme();

  if (props.circle === true) {
    return {
      borderColor: palette.skeColor.darkBlue,
      backgroundColor: palette.skeColor.darkBlue
    };
  } else {
    return {
      borderColor: palette.skeColor.darkBlue,
      backgroundColor: palette.skeColor.darkBlue
    };
  }
}

function getIconSize(props) {
  switch (props.buttonType) {
    case 'small':
      return {
        fontSize: FontSizes.mediumPlus
      };
    case 'medium':
      return {
        fontSize: FontSizes.largePlus
      };
    case 'large':
      return {
        fontSize: FontSizes.xxLarge
      };
    case 'xLarge':
      return {
        fontSize: FontSizes.superLarge
      };
    default:
      return {
        fontSize: FontSizes.icon
      };
  }
}

export var getClassNames = function getClassNames(props) {
  const { palette } = getTheme();
  const inset = -4;
  const radius = 0;

  return mergeStyles([
    getFocusStyle({ palette }, inset, 'relative', radius),
    {
      displayName: `ske-icon${props.buttonType}-button`,
      selectors: {
        '&.ms-Button--icon': {
          width: '40px',
          height: '40px',
          borderRadius: '50%',
          padding: 0,
          background: 'none',
          color: palette.skeColor.blue,
          ...getTypeColor(props),
          ...getCircleStyles(props)
        },
        '&.ms-Button--icon i': {
          ...getIconSize(props)
        },
        '&.ms-Button--icon:hover, &.ms-Button--icon:focus': {
          border: '3px solid',
          borderColor: palette.skeColor.blue,
          transition: 'background-color 0.3s',
          ...getCircleHoverStyles(props)
        },
        '.ms-Fabric.is-focusVisible.is-focusVisible &:focus:after': {
          borderWidth: '0px'
        },
        '&.ms-Button--icon:active': {
          color: palette.white,
          ...getCircleFocusrStyles(props)
        },
        '&.ms-Button--icon:disabled': {
          ...getCircleStyles(props),
          color: palette.skeLysGraa,
          borderColor: palette.skeColor.lightGrey
        },
        '&.ms-Button--icon:disabled i': {
          color: palette.skeColor.lightGrey
        }
      }
    }
  ]);
};
