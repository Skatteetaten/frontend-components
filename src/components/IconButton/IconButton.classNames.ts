import { mergeStyles } from '@uifabric/merge-styles';
import { getTheme } from '@uifabric/styling';
import { FontSizes } from '../utils/fonts';
import { getFocusStyle } from '../utils/getFocusStyle';
import { PaletteProps } from '..';
// @ts-ignore TODO
function getTypeColor(props) {
  switch (props.type) {
    case 'small':
      return {
        width: '35px',
        height: '35px'
      };
    case 'medium':
      return {
        width: '40px',
        height: '40px'
      };
    case 'large':
      return {
        width: '50px',
        height: '50px'
      };
    case 'xLarge':
      return {
        width: '60px',
        height: '60px'
      };
    default:
      return {};
  }
}
// @ts-ignore TODO
function getCircleStyles(props) {
  const palette = getTheme().palette as PaletteProps;

  if (props.circle === true) {
    return {
      border: 'solid',
      borderWidth: '3px',
      borderColor: palette.skeColor.blue
    };
  } else {
    return {
      borderWidth: 0
    };
  }
}
// @ts-ignore TODO
function getCircleHoverStyles(props) {
  const palette = getTheme().palette as PaletteProps;

  if (props.circle === true) {
    return {
      borderWidth: '3px',
      background: palette.skeColor.lightBlue,
      color: palette.skeColor.blue
    };
  } else {
    return {
      color: palette.skeColor.blue,
      borderWidth: '3px',
      borderColor: palette.skeColor.blue,
      background: palette.skeColor.lightBlue
    };
  }
}
// @ts-ignore TODO
function getCircleFocusrStyles(props) {
  const palette = getTheme().palette as PaletteProps;

  if (props.circle === true) {
    return {
      borderColor: palette.skeColor.darkBlue,
      backgroundColor: palette.skeColor.darkBlue
    };
  } else {
    return {
      borderColor: palette.skeColor.darkBlue,
      backgroundColor: palette.skeColor.darkBlue
    };
  }
}
// @ts-ignore TODO
function getIconSize(props) {
  switch (props.type) {
    case 'small':
      return {
        fontSize: FontSizes.mediumPlus
      };
    case 'medium':
      return {
        fontSize: FontSizes.largePlus
      };
    case 'large':
      return {
        fontSize: FontSizes.xxLarge
      };
    case 'xLarge':
      return {
        fontSize: FontSizes.superLarge
      };
    default:
      return {
        fontSize: FontSizes.icon
      };
  }
}
// @ts-ignore TODO
export var getClassNames = function getClassNames(props) {
  const palette = getTheme().palette as PaletteProps;
  const inset = -4;
  const radius = 0;

  return mergeStyles([
    // @ts-ignore TODO
    getFocusStyle({ palette }, inset, 'relative', radius),
    {
      displayName: `ske-icon${props.type}-button`,
      selectors: {
        '&.ms-Button--icon': {
          width: '40px',
          height: '40px',
          borderRadius: '50%',
          padding: 0,
          background: 'none',
          color: palette.skeColor.blue,
          ...getTypeColor(props),
          ...getCircleStyles(props)
        },
        '&.ms-Button--icon i': {
          ...getIconSize(props)
        },
        '&.ms-Button--icon:hover, &.ms-Button--icon:focus': {
          border: '3px solid',
          borderColor: palette.skeColor.blue,
          transition: 'background-color 0.3s',
          ...getCircleHoverStyles(props)
        },
        '.ms-Fabric.is-focusVisible.is-focusVisible &:focus:after': {
          borderWidth: '0px'
        },
        '&.ms-Button--icon:active': {
          color: palette.white,
          ...getCircleFocusrStyles(props)
        },
        '&.ms-Button--icon:disabled': {
          ...getCircleStyles(props),
          // @ts-ignore TODO
          color: palette.skeLysGraa,
          borderColor: palette.skeColor.lightGrey
        },
        '&.ms-Button--icon:disabled i': {
          color: palette.skeColor.lightGrey
        }
      }
    }
  ]);
};
