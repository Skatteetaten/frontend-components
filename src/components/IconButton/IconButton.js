import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { IconButton as FabricIconButton } from 'office-ui-fabric-react/lib-commonjs/Button';

import { getClassNames } from './IconButton.classNames';

/**
 * @visibleName IconButton (Ikonknapp)
 */
export default class IconButton extends React.PureComponent {
  static propTypes = {
    /** Gjør knappen til hovedknapp med fylt farge */
    buttonType: PropTypes.oneOf([
      'default',
      'small',
      'medium',
      'large',
      'xLarge'
    ]),
    /** Om knappen er inaktiv*/
    disabled: PropTypes.bool,
    /** Om sirkel skal vises eller ikke. Sirkel vil typisk benyttes når ikon fra material-design ikke har egen sirkel  */
    circle: PropTypes.bool,
    /** Ikon som skal vises foran teksten på knappen */
    icon: PropTypes.string,
    /** Beskrivelse av hva knappen gjør (f.eks. til skjermlesere) */
    title: PropTypes.string,
    /** Callback for klikk på knappen */
    onClick: PropTypes.func,
    className: PropTypes.string,
    /** Tekst alternativ til ikonet. Hvis bildet ikke har noen betydning vil attributten vise " " som default */
    alt: PropTypes.string
  };
  static defaultProps = {
    buttonType: 'default',
    circle: false,
    icon: undefined,
    title: undefined,
    disabled: undefined,
    onClick: undefined,
    alt: ' '
  };

  render() {
    const {
      buttonType,
      icon,
      title,
      disabled,
      onClick,
      className,
      alt,
      ...props
    } = this.props;
    return (
      <FabricIconButton
        {...props}
        className={classnames(getClassNames(this.props), className)}
        buttonType={buttonType}
        title={title}
        aria-label={title}
        role="button"
        iconProps={{ iconName: icon }}
        disabled={disabled}
        onClick={onClick}
        alt={alt}
      />
    );
  }
}
