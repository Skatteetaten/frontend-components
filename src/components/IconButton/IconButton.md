** Ikonknapper brukes til mindre fremtredende funksjoner, når man har lite plass eller når funksjonen er innlysende og man ikke trenger å kommunisere den tekstlig **

```js
import IconButton from '@skatteetaten/frontend-components/IconButton';

<div className="ExampleSpacing8">
  <IconButton
    alt={'Sky med sirkel rundt'}
    title="Skriv ut"
    circle
    type="large"
    icon="Print"
  />
  <IconButton
    alt={'Sky med sirkel rundt'}
    title="Skriv ut"
    type="large"
    icon="Print"
  />
  <br />
  <IconButton
    alt={'Sky med sirkel rundt'}
    title="Skriv ut"
    circle
    type="default"
    icon="Print"
  />
  <IconButton
    alt={'Sky med sirkel rundt'}
    title="Skriv ut"
    type="default"
    icon="Print"
  />
</div>;
```

```js noeditor
import Accordion from '@skatteetaten/frontend-components/Accordion';
import AccordionItem from '@skatteetaten/frontend-components/Accordion/AccordionItem';

<Accordion>
  <AccordionItem
    isOpen
    toggleContent
    toggleButtonText={'Bruk'}
    stepId={'step-1-1'}
  >
    <p>
      Knapper bør i størst mulig grad ha den beskrivende teksten synlig, men
      ikonknapper kan brukes i situasjoner der man ikke har anleding til å vise
      den. Det skal alltid være en tekst som kommer ved mouseover som forklarer
      hva ikonet gjør.
    </p>
    <p>Ikonknapper bør brukes med varsomhet i publikumsløsninger.</p>
    <p>
      Ikonknapper kan være små eller store, med eller uten sirkel. Sirkelen
      benyttes for å fremheve ikonknappen.
    </p>
  </AccordionItem>
  <AccordionItem
    toggleContent
    toggleButtonText={'Universell utforming'}
    stepId={'step-1-2'}
  >
    <ul>
      <li>Tittel på knappen bør beskrive dens funksjon.</li>
      <li>
        Husk å bruke gi ikonknappen en label slik at skjermlesere kan lese opp
        funksjonen
      </li>
    </ul>
  </AccordionItem>
  <AccordionItem
    toggleContent
    toggleButtonText={'Tekniske detaljer'}
    stepId={'step-1-3'}
  >
    <p>
      <a
        href="https://developer.microsoft.com/en-us/fabric#/components/button#Implementation"
        target="_blank"
      >
        Se flere tilgjengelige props i Fabric dokumentasjonen
      </a>
    </p>
  </AccordionItem>
</Accordion>;
```
