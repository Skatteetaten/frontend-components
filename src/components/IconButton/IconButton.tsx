import classnames from 'classnames';
import * as React from 'react';

import {
  IconButton as FabricIconButton,
  IButtonProps
} from 'office-ui-fabric-react/lib-commonjs/Button';

import { getClassNames } from './IconButton.classNames';

interface IconButtonProps extends IButtonProps {
  /** Om sirkel skal vises eller ikke. Sirkel vil typisk benyttes når ikon fra material-design ikke har egen sirkel  */
  circle?: boolean;
  /** Ikon som skal vises foran teksten på knappen */
  icon?: string;
  /** Gjør knappen til hovedknapp med fylt farge */
  type?: 'default' | 'small' | 'medium' | 'large' | 'xLarge';
}
/**
 * @visibleName IconButton (Ikonknapp)
 */
export default class IconButton extends React.PureComponent<
  IconButtonProps,
  {}
> {
  static defaultProps = {
    alt: ' ',
    circle: false,
    disabled: undefined,
    icon: undefined,
    onClick: undefined,
    title: undefined,
    type: 'default'
  };

  render() {
    const { icon, className, ...props } = this.props;
    return (
      <FabricIconButton
        {...props}
        className={classnames(getClassNames(this.props), className)}
        role="button"
        iconProps={{ iconName: icon }}
      />
    );
  }
}
