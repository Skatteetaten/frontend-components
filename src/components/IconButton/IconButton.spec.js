import React from 'react';
import { shallow, mount } from 'enzyme';
import IconButton from './IconButton';
import { matches } from './../utils/test-utils';

function oppsettShallow(props) {
  return shallow(<IconButton {...props} />);
}

function oppsettFullDOM(props) {
  return mount(<IconButton {...props} />);
}

describe('IconButton komponent', () => {
  it('matcher snapshot', () => {
    matches(
      <IconButton title="Lagre" circle={true} type="default" icon="Save" />
    );
  });

  it('rendrer IconButton med default props', () => {
    const wrapper = oppsettShallow({});

    expect(wrapper.prop('circle')).toEqual(false);
    expect(wrapper.prop('type')).toEqual('default');
    expect(wrapper.prop('icon')).toEqual(undefined);
    expect(wrapper.prop('title')).toEqual(undefined);
    expect(wrapper.prop('disabled')).toEqual(undefined);
    expect(wrapper.prop('onClick')).toEqual(undefined);
  });

  it('rendrer IconButton med riktige props', () => {
    const wrapper = oppsettFullDOM({
      title: 'Lagre',
      circle: true,
      type: 'large',
      icon: 'Save'
    });

    expect(wrapper.prop('title')).toEqual('Lagre');
    expect(wrapper.prop('circle')).toEqual(true);
    expect(wrapper.prop('type')).toEqual('large');
    const icon = wrapper.find('.ms-Icon');
    expect(icon.prop('data-icon-name')).toEqual('Save');
  });
});
