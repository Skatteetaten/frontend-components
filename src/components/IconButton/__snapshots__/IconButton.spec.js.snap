// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton komponent matcher snapshot 1`] = `
<button
  className="ms-Button ms-Button--icon ske-icondefault-button-41"
  data-is-focusable={true}
  onClick={[Function]}
  onKeyDown={[Function]}
  onKeyPress={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  role="button"
  title="Lagre"
  type="default"
>
  <span
    className="ms-Button-flexContainer flexContainer-42"
    data-automationid="splitbuttonprimary"
  >
    <i
      aria-hidden={true}
      className="ms-Icon root-33 css-49 ms-Button-icon icon-44"
      data-icon-name="Save"
      role="presentation"
    >
      ï†“
    </i>
  </span>
</button>
`;
